athena-gcs/src/main/java/com/amazonaws/athena/connectors/gcs/GcsRecordHandler.java

```diff
@@ -0,0 +1,202 @@
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.util.VisibleForTesting;
+import org.apache.arrow.vector.FieldVector;
+import org.apache.arrow.vector.VectorSchemaRoot;
+import org.apache.arrow.vector.ipc.ArrowReader;
+import org.apache.arrow.vector.types.Types;
+import org.apache.arrow.vector.types.pojo.Field;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+
+import static com.amazonaws.athena.connectors.gcs.GcsConstants.CLASSIFICATION_GLUE_TABLE_PARAM;
+import static com.amazonaws.athena.connectors.gcs.GcsThrottlingExceptionFilter.EXCEPTION_FILTER;
+import static com.amazonaws.athena.connectors.gcs.GcsUtil.createUri;
+
+public class GcsRecordHandler
+        extends RecordHandler
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(GcsRecordHandler.class);
+
+    private static final String SOURCE_TYPE = "gcs";
+    public static final int BATCH_SIZE = 32768;
+    private BufferAllocator allocator;
+
+    // to handle back-pressure during API invocation to GCS
+    ThrottlingInvoker invoker = ThrottlingInvoker.newDefaultBuilder(EXCEPTION_FILTER).build();
+
+    public GcsRecordHandler(BufferAllocator allocator)
+    {
+        this(AmazonS3ClientBuilder.defaultClient(),
+                AWSSecretsManagerClientBuilder.defaultClient(),
+                AmazonAthenaClientBuilder.defaultClient());
+        this.allocator = allocator;
+    }
+
+    /**
+     * Constructor that provides access to S3, secret manager and athena
+     * <p>
+     * @param amazonS3       An instance of AmazonS3
+     * @param secretsManager An instance of AWSSecretsManager
+     * @param amazonAthena   An instance of AmazonAthena
+     */
+    @VisibleForTesting
+    protected GcsRecordHandler(AmazonS3 amazonS3, AWSSecretsManager secretsManager, AmazonAthena amazonAthena)
+    {
+        super(amazonS3, secretsManager, amazonAthena, SOURCE_TYPE);
+    }
+
+    /**
+     * Used to read the row data associated with the provided Split.
+     *
+     * @param spiller            A BlockSpiller that should be used to write the row data associated with this Split.
+     *                           The BlockSpiller automatically handles chunking the response, encrypting, and spilling to S3.
+     * @param recordsRequest     Details of the read request, including:
+     *                           1. The Split - it contains information
+     *                           2. The Catalog, Database, and Table the read request is for.
+     *                           3. The filtering predicate (if any)
+     *                           4. The columns required for projection.
+     * @param queryStatusChecker A QueryStatusChecker that you can use to stop doing work for a query that has already terminated
+     */
+    @Override
+    protected void readWithConstraint(BlockSpiller spiller, ReadRecordsRequest recordsRequest,
```

Comment:
why are we not pushing down any constraints?

Reply:
It is not pushed down the source, so that's why we are not pushing down any constraints. It is already discussed in the design phase.

Reply:
Discussed this with Burhan internally. Closing this part out for now.